sandbox/apparmor/apparmor.go

```diff
@@ -48,6 +49,401 @@ func ValidateNoAppArmorRegexp(s string) error {
+		pattChar := excludePatterns[0][i]
+		for _, patt := range excludePatterns[1:] {
+			if patt[i] != pattChar {
+				break commonSubStrLoop
+			}
+		}
+		endCommonPrefix = i + 1
+	}
+
+	// for keeping track of the last rule for the common case before we get to
+	// the overlapping cases
+	var lastCommonPrefix string
+
+	// iterate up to the longest string we have - saving the iteration variable
+	// for the case when we get past the common substrings and do a separate
+	// kind of loop below, mainly just for readability to avoid nesting things
+	// too deeply
+	var i int
+charLoop:
+	for i = 0; i < longestStrLen; i++ {
+		switch {
+		case i < endCommonPrefix:
+			// then generate common simple rules
+			c := excludePatterns[0][i]
+			switch c {
+			case '*':
+				// skip this element, this is the wildcard
+				continue
+			case '/':
+				// check if this is the first character or if the previous
+				// element was a "*" in which case we skip this one too
+				if i == 0 {
+					// this is for the case where the only common character is
+					// actually the first "/" character
+					lastCommonPrefix = string(c)
+					continue
+				}
+				if excludePatterns[0][i-1] == '*' {
+					// this is for the case where the last common characters are
+					// "*/" if there are other common characters after this, it
+					// will get overwritten on the next iteration
+					lastCommonPrefix = lastCommonPrefix + "*/"
+					continue
+				}
+			}
+			// snippet up to this character so far - again since this is the
+			// common prefix we can just use the first pattern
+			snippet := string(excludePatterns[0][:i])
+			res := fmt.Sprintf("%s%s[^%c]**%s\n", opts.Prefix, snippet, c, opts.Suffix)
+
+			builder.WriteString(res)
+
+			// save this as the most recent common sub string pattern
+			lastCommonPrefix = snippet + string(c)
+		case i == endCommonPrefix:
+			// this is the last bit of the common substring, so get all the
+			// characters for each pattern and just make a negative match for
+			// all of those characters
+
+			// note we use a map because there could be duplicate characters,
+			// for example we could have 3 strings like "abcd1", "abcd2", "abceee",
+			// where after the first common characters among the 3, there is
+			// another common character with first 2 exclude patterns not in the
+			// third
+			chars := map[rune]bool{}
+			for _, pattern := range excludePatterns {
+				if i >= len(pattern) {
+					continue
+				}
+				chars[pattern[i]] = true
+			}
+			if len(chars) == 0 {
+				// shouldn't happen in practice, we check for duplicates above
+				// but just be on the safe side
+				return "", fmt.Errorf("internal erorr: all excluded patterns are the same")
+			}
+
+			negGroup := []rune("[^")
+			for c := range chars {
+				negGroup = append(negGroup, c)
+			}
+			// make sure the runs are sorted for consistency
+			sort.Slice(negGroup, func(i, j int) bool {
+				return negGroup[i] < negGroup[j]
+			})
```

Comment:
Try adding some capital letters or numbers in the unit tests, and you'll see that the generated patters are not correct :-)

(you are also sorting the `[` and the `^`, which sort before the lowercase letters but after the numbers and the uppercase letters)

Reply:
Indeed fixed :slightly_smiling_face: